src/core/types/message.rs

```diff
@@ -4,17 +4,196 @@ use crate::core::types::{
     SuccessfulPayment, User, Venue, Video, VideoNote, Voice,
 };
 
-#[derive(Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+#[derive(Debug, Deserialize, Eq, Hash, PartialEq)]
 pub struct Message {
     #[serde(rename = "message_id")]
-    pub id: i32,
+    pub id: i64,
     pub date: i32,
-    pub chat: Chat,
+    pub chat: Box<Chat>,
     #[serde(flatten)]
-    pub type_: MessageType,
+    pub message_kind: MessageKind,
 }
 
-#[derive(Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
-#[serde(rename_all = "snake_case")]
+#[derive(Debug, Deserialize, Eq, Hash, PartialEq)]
 #[serde(untagged)]
-pub enum MessageType {}
+pub enum MessageKind {
+    IncomingMessage {
+        #[serde(flatten)]
+        forward_kind: ForwardKind,
+        edit_date: Option<i32>,
+        #[serde(flatten)]
+        media_kind: MediaKind,
+        reply_markup: Option<InlineKeyboardMarkup>,
+    },
+    NewChatMembers {
+        new_chat_members: Vec<User>,
+    },
+    LeftChatMember {
+        left_chat_member: User,
+    },
+    NewChatTitle {
+        new_chat_title: String,
+    },
+    NewChatPhoto {
+        new_chat_photo: Vec<PhotoSize>,
+    },
+    DeleteChatPhoto {
+        delete_chat_photo: bool,
+    },
+    GroupChatCreated {
+        group_chat_created: bool,
+    },
+    SupergroupChatCreated {
+        supergroup_chat_created: bool,
+    },
+    ChannelChatCreated {
+        channel_chat_created: bool,
+    },
+    Migrate {
+        migrate_to_chat_id: i64,
+        migrate_from_chat_id: i64,
+    },
+    PinnedMessage {
+        pinned_message: Box<Message>,
+    },
+    Invoice {
+        invoice: Invoice,
+    },
+    SuccessfulPayment {
+        successful_payment: SuccessfulPayment,
+    },
+    ConnectedWebsite {
+        connected_website: String,
+    },
+    PassportData {
+        passport_data: PassportData,
+    },
+}
+
+#[derive(Debug, Deserialize, Eq, Hash, PartialEq)]
+#[serde(untagged)]
+pub enum ForwardKind {
+    ChannelForward {
+        #[serde(rename = "forward_date")]
+        date: i32,
+        #[serde(rename = "forward_from_chat")]
+        chat: Box<Chat>,
+        #[serde(rename = "forward_from_message_id")]
+        message_id: i64,
+        #[serde(rename = "forward_signature")]
+        signature: Option<String>,
+    },
+    NonChannelForward {
+        #[serde(rename = "forward_date")]
+        date: i32,
+        #[serde(flatten)]
+        from: ForwardedFrom,
+    },
+    Origin {
+        reply_to_message: Option<Box<Message>>,
+    },
+}
+
+#[derive(Debug, Deserialize, Eq, Hash, PartialEq)]
+#[serde(untagged)]
+pub enum ForwardedFrom {
```

Comment:
Use tuple structs:
```rust
#[derive(Debug, Deserialize, Eq, Hash, PartialEq)]
pub enum ForwardedFrom {
    #[serde(rename = "forward_from")]
    User(User),
    #[serde(rename = "forward_sender_name")]
    SenderName(String),
}
```

Reply:
When you don't `#[serde(untagged)]` enum, struct-like variant is de/serialized as `{"variant_name": {"field": "value", ...}}`. If I use tuple-like variant, won't it de/serialize as `{"variant_name": ["value", ...]}`?

Reply:
I have tested it already and it works as expected. This test passes:
```rust
#[derive(Debug, Deserialize, Eq, Hash, PartialEq)]
pub enum ForwardedFrom {
    #[serde(rename = "forward_from")]
    User(User),
    #[serde(rename = "forward_sender_name")]
    SenderName(String),
}

#[test]
fn test() {
    let expected = ForwardedFrom::SenderName(String::from("name"));
    let actual: ForwardedFrom = serde_json::from_str(r#"{"forward_sender_name":"name"}"#).unwrap();

    assert_eq!(expected, actual);
}
```

Reply:
Btw, that's exactly the thing why I'll probably try to `impl Deserialize for Message` by hand. I want this syntax as well, but I don't understand how do I implement this with `serde(...)`

Reply:
Much wow, okay, I fixed